src/WebJobs.Script.Analyzers/AssemblyCache.cs

```diff
@@ -0,0 +1,190 @@
+
+        internal JobHostMetadataProvider Tooling => _tooling;
+        private int _projectCount;
+
+        // $$$ This can get invoked multiple times concurrently
+        // This will get called on every compilation.
+        // So return early on subsequent initializations.
+        internal void Build(Compilation compilation)
+        {
+            Register();
+
+            int count;
+            lock (this)
+            {
+                // If project references have changed, then reanalyze to pick up new dependencies.
+                var refs = compilation.References.OfType<PortableExecutableReference>().ToArray();
+                count = refs.Length;
+                if ((count == _projectCount) && (_tooling != null))
+                {
+                    return; // already initialized.
+                }
+
+                // Even for netStandard/.core projects, this will still be a flattened list of the full transitive closure of dependencies.
+                foreach (var asm in compilation.References.OfType<PortableExecutableReference>())
+                {
+                    var dispName = asm.Display; // For .net core, the displayname can be the full path
+                    var path = asm.FilePath;
+
+                    _map[dispName] = path;
+                }
+
+                // Builtins
+                _mapRef["mscorlib"] = typeof(object).Assembly;
+                _mapRef[WebJobsAssemblyName] = typeof(Microsoft.Azure.WebJobs.FunctionNameAttribute).Assembly;
+                _mapRef[WebJobsHostAssemblyName] = typeof(Microsoft.Azure.WebJobs.JobHost).Assembly;
+
+                // JSON.Net?
+            }
+
+            // Produce tooling object
+            var webjobsStartups = new List<Type>();
+            foreach (var path in _map.Values)
+            {
+                // We don't want to load and reflect over every dll.
+                // By convention, restrict based on filenames.
+                var filename = Path.GetFileName(path);
+                // TODO: I assume this is so general to avoid missing custom extensions, but can it be tightened up for performance?
+                if (!filename.ToLowerInvariant().Contains("extension"))
+                {
+                    continue;
+                }
+                if (path.Contains(@"\ref\"))    // Skip reference assemblies.
+                {
+                    continue; 
+                }
+
+                Assembly assembly;
+                try
+                {
+                    // See GetNuGetPackagesPath for details
+                    // Script runtime is already setup with assembly resolution hooks, so use LoadFrom
+                    assembly = Assembly.LoadFrom(path);
+
+                    string asmName = new AssemblyName(assembly.FullName).Name;
+                    _mapRef[asmName] = assembly;
+
+                    var test = assembly.GetCustomAttributes<WebJobsStartupAttribute>().Select(a => a.WebJobsStartupType);
```

Comment:
Not sure that I understand why the variable is named test? It looks like you're selecting a startup type? I might be missing some context

Reply:
I think it means "I'm testing to see if this assembly has a `WebJobsStartupAttribute`" but I agree the name is a little indirect. I'll change it to `startupAttr`.